File: test/features/rankings/data/rankings_provider_test.dart
import 'package:clashkingapp/features/rankings/data/rankings_provider.dart';
import 'package:clashkingapp/features/rankings/data/rankings_service.dart';
import 'package:clashkingapp/features/rankings/models/ranking_models.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('every board and filter change issues a new request', () async {
    final service = _RecordingRankingsService();
    final provider = RankingsProvider(
      service: service,
      leagueOptions: const [
        RankingLeagueOption.legendOne,
        RankingLeagueOption(
          id: 105000033,
          name: 'Legend II',
          iconUrl: 'legend-2',
        ),
      ],
      clock: () => DateTime(2026, 7, 20),
    );

    await provider.initialize();
    expect(service.queries, hasLength(1));
    expect(service.queries.last.board, RankingBoard.playerHome);
    expect(service.queries.last.location.apiPath, 'global');

    await provider.selectLocation(service.locations.last);
    expect(service.queries.last.location.name, 'United States');

    await provider.selectBoard(RankingBoard.playerTownHall);
    await provider.selectTownHall(17);
    await provider.selectPeriod(RankingPeriod.history);
    await provider.selectHistoryDate(DateTime(2026, 7, 18));
    expect(service.queries.last.board, RankingBoard.playerTownHall);
    expect(service.queries.last.townHallLevel, 17);
    expect(service.queries.last.period, RankingPeriod.history);
    expect(service.queries.last.historyDate, DateTime(2026, 7, 18));

    await provider.selectBoard(RankingBoard.playerRanked);
    await provider.selectLeague(provider.leagueOptions.last);
    expect(service.queries.last.leagueTier.id, 105000033);

    await provider.selectAudience(RankingAudience.clans);
    await provider.selectBoard(RankingBoard.clanDonations);
    expect(service.queries.last.board, RankingBoard.clanDonations);
    expect(service.queries.last.period, RankingPeriod.current);
    expect(service.queries.length, 10);
  });

  test('surfaces a clear empty history result without an error', () async {
    final service = _RecordingRankingsService(empty: true);
    final provider = RankingsProvider(
      service: service,
      leagueOptions: const [RankingLeagueOption.legendOne],
      clock: () => DateTime(2026, 7, 20),
    );

    await provider.initialize();
    await provider.selectPeriod(RankingPeriod.history);

    expect(provider.error, isNull);
    expect(provider.result?.entries, isEmpty);
    expect(provider.isLoading, isFalse);
  });
}

class _RecordingRankingsService extends RankingsService {
  _RecordingRankingsService({this.empty = false});

  final bool empty;
  final queries = <RankingQuery>[];
  final locations = const [
    RankingLocation.worldwide(),
    RankingLocation(id: 32000006, name: 'International', isCountry: false),
    RankingLocation(
      id: 32000007,
      name: 'United States',
      isCountry: true,
      countryCode: 'US',
    ),
  ];

  @override
  Future<List<RankingLocation>> fetchLocations() async => locations;

  @override
  Future<RankingResult> fetchRankings(RankingQuery query) async {
    queries.add(query);
    return RankingResult(
      entries: empty
          ? const []
          : [
              RankingEntry(
                audience: query.board.audience,
                rank: 1,
                previousRank: 2,
                tag: query.board.isClan ? '#CLAN' : '#PLAYER',
                name: query.board.isClan ? 'Clan One' : 'Player One',
                subtitle: query.board.isClan ? '#CLAN' : '#PLAYER',
                score: 6000,
                imageUrl: query.board.iconUrl,
                metricImageUrl: query.board.iconUrl,
                townHallLevel: 18,
              ),
            ],
      source: query.board.source,
      limit: query.board.source == RankingSource.official ? 200 : 500,
    );
  }
}

File: expo/src/features/rankings/data/rankings-provider.test.ts
import {
  gameDataState,
  resetGameDataStateForTesting,
} from '../../../core/game-data/game-data-state';
import {
  RankingAudience,
  RankingBoard,
  RankingEntry,
  RankingLeagueOption,
  RankingLocation,
  RankingPeriod,
  RankingResult,
  type RankingQuery,
} from '../models';
import { rankingLeagueOptionsFromGameData, RankingsProvider } from './rankings-provider';
import { RankingsRequestException, type RankingsServiceContract } from './rankings-service';

afterEach(resetGameDataStateForTesting);

test('Capital rejects non-Monday calendar dates and normalizes legacy archive selections', async () => {
  const service = new RecordingService();
  const provider = new RankingsProvider(service, { clock: () => new Date(2026, 8, 23) });
  await provider.openBoard(RankingBoard.clanCapital, false);
  await provider.selectDate(new Date(2026, 8, 22));
  expect(service.queries).toHaveLength(0);
  await provider.selectDate(new Date(2026, 8, 21));
  expect(service.queries.at(-1)).toMatchObject({
    period: RankingPeriod.history,
    historyDate: new Date(2026, 8, 21),
  });
  await provider.selectDate(provider.today);
  expect(service.queries.at(-1)?.period).toBe(RankingPeriod.current);
  provider.period = RankingPeriod.history;
  provider.historyDate = new Date(2026, 8, 20);
  await provider.reload();
  expect(service.queries.at(-1)?.historyDate).toEqual(new Date(2026, 8, 14));
});

test('builds every game-data tier except Legend League 1 in descending ID order', () => {
  gameDataState.playerLeagueData.leagues = {
    'Legend 1': { _id: 105000036, name: 'Legend League' },
    'Legend 2': { _id: 105000035, name: 'Legend League' },
    'Legend 3': { _id: 105000034, name: 'Legend League' },
    'Pekka 23': { _id: 105000033, name: 'P.E.K.K.A League 23' },
    invalid: { _id: '105000032oops', name: 'Invalid' },
  };
  const options = rankingLeagueOptionsFromGameData();
  expect(options.map((option) => option.id)).toEqual([105000035, 105000034, 105000033]);
  expect(options.map((option) => option.name)).toEqual([
    'Legend League 2',
    'Legend League 3',
    'P.E.K.K.A League 23',
  ]);
});

test('preserves board/filter behavior and matching village board across audiences', async () => {
  const service = new RecordingService();
  const provider = new RankingsProvider(service, {
    leagueOptions: [RankingLeagueOption.legendTwo, RankingLeagueOption.legendThree],
    clock: () => new Date(2026, 6, 20, 12),
  });
  await provider.initialize();
  expect(provider.historyDate).toEqual(new Date(2026, 6, 19));
  await provider.selectBoard(RankingBoard.playerBuilder);
  await provider.selectAudience(RankingAudience.clans);
  expect(provider.board).toBe(RankingBoard.clanBuilder);
  await provider.selectBoard(RankingBoard.clanDonations);
  expect(provider.location.name).toBe('United States');
  await provider.selectPeriod(RankingPeriod.history);
  expect(provider.period).toBe(RankingPeriod.current);
  await provider.selectAudience(RankingAudience.players);
  expect(provider.board).toBe(RankingBoard.playerHome);
});

test('maps daily navigation and calendar selection between live and history with one request each', async () => {
  const service = new RecordingService();
  const provider = new RankingsProvider(service, {
    leagueOptions: [RankingLeagueOption.legendTwo],
    clock: () => new Date(2026, 8, 23, 17),
  });
  await provider.initialize();
  const baseline = service.queries.length;

  await provider.selectDate(new Date(2026, 8, 22));
  expect(provider.period).toBe(RankingPeriod.history);
  expect(provider.historyDate).toEqual(new Date(2026, 8, 22));
  expect(service.queries).toHaveLength(baseline + 1);
  expect(service.queries.at(-1)).toMatchObject({
    period: RankingPeriod.history,
    historyDate: new Date(2026, 8, 22),
  });

  await provider.selectDate(new Date(2026, 8, 23));
  expect(provider.period).toBe(RankingPeriod.current);
  expect(service.queries).toHaveLength(baseline + 2);
  expect(service.queries.at(-1)).toMatchObject({ period: RankingPeriod.current });

  await provider.selectDate(new Date(2026, 8, 23));
  await provider.selectDate(new Date(2026, 8, 24));
  await provider.selectDate(new Date(2023, 8, 22));
  expect(service.queries).toHaveLength(baseline + 2);
});

test('does not offer daily selection for current-only boards', async () => {
  const service = new RecordingService();
  const provider = new RankingsProvider(service, { clock: () => new Date(2026, 8, 23) });
  await provider.initialize();
  await provider.openBoard(RankingBoard.playerTownHall);
  const baseline = service.queries.length;
  await provider.selectDate(new Date(2026, 8, 22));
  expect(provider.period).toBe(RankingPeriod.current);
  expect(service.queries).toHaveLength(baseline);
});

test('treats explicit and message-based not-found failures as empty results', async () => {
  for (const error of [
    new RankingsRequestException(404),
    new Error('No rankings found for selected filters.'),
  ]) {
    const provider = new RankingsProvider(new FailingService(error), {
      leagueOptions: [RankingLeagueOption.legendTwo],
    });
    await provider.initialize();
    expect(provider.error).toBeNull();
    expect(provider.result?.entries).toEqual([]);
    expect(provider.isLoading).toBe(false);
  }
});

test('ignores an older response that completes after a newer reload', async () => {
  const deferred: {
    query: RankingQuery;
    resolve: (result: RankingResult) => void;
  }[] = [];
  const service: RankingsServiceContract = {
    fetchLocations: async () => [RankingLocation.worldwide()],
    fetchRankings: (query) =>
      new Promise((resolve) => {
        deferred.push({ query, resolve });
      }),
  };
  const provider = new RankingsProvider(service, {
    leagueOptions: [RankingLeagueOption.legendTwo],
  });
  const older = provider.reload();
  provider.townHallLevel = 17;
  const newer = provider.reload();
  deferred[1]?.resolve(resultFor(deferred[1].query, '#NEW'));
  await newer;
  deferred[0]?.resolve(resultFor(deferred[0].query, '#OLD'));
  await older;
  expect(provider.result?.entries[0]?.tag).toBe('#NEW');
  expect(provider.isLoading).toBe(false);
});

test('loads locations before the initial location-only leaderboard', async () => {
  const service = new RecordingService();
  const provider = new RankingsProvider(service, {
    leagueOptions: [RankingLeagueOption.legendTwo],
  });
  provider.audience = RankingAudience.clans;
  provider.clanBoard = RankingBoard.clanDonations;

  await provider.initialize();

  expect(service.queries).toHaveLength(1);
  expect(service.queries[0]).toMatchObject({
    board: RankingBoard.clanDonations,
    location: expect.objectContaining({ id: 32000007 }),
  });
});

class RecordingService implements RankingsServiceContract {
  readonly queries: RankingQuery[] = [];
  async fetchLocations() {
    return [
      RankingLocation.worldwide(),
      new RankingLocation(32000007, 'United States', true, 'US'),
    ];
  }
  async fetchRankings(query: RankingQuery) {
    this.queries.push(query);
    return resultFor(query, query.board.isClan ? '#CLAN' : '#PLAYER');
  }
}

class FailingService implements RankingsServiceContract {
  constructor(private readonly failure: unknown) {}
  async fetchLocations() {
    return [RankingLocation.worldwide()];
  }
  async fetchRankings(_query: RankingQuery): Promise<RankingResult> {
    throw this.failure;
  }
}

function resultFor(query: RankingQuery, tag: string): RankingResult {
  return new RankingResult(
    [
      new RankingEntry(
        query.board.audience,
        1,
        2,
        tag,
        tag,
        '',
        6000,
        query.board.iconUrl,
        query.board.iconUrl,
        18,
      ),
    ],
    query.board.source,
    query.board.source === 'official' ? 200 : 500,
  );
}
